sonosco/training/helpers.py

```diff
@@ -0,0 +1,143 @@
+import logging
+import sys
+import os.path as path
+import numpy as np
+import torch
+
+from collections import defaultdict
+from .abstract_callback import AbstractCallback
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class HistoryRecorder(AbstractCallback):
+    """ Records all losses and metrics during training. """
+
+    def __init__(self, epoch_steps):
+        self.history = defaultdict(list)
+        self._epoch_steps = epoch_steps
+
+    def __call__(self, epoch, step, performance_measures, context):
+
+        if step % self._epoch_steps == 0:  # only record at end of epoch
+            return
+
+        for key, value in performance_measures.items():
+            if type(value) == torch.Tensor:
+                value = value.item()
+            self.history[key].append(value)
+
+
+class ModelCheckpoint(AbstractCallback):
```

Comment:
To a separate file

Reply:
I would argue against putting everything in a separate module. That doesn't bring additional benefit but just clutters the whole codebase with many-many small files.